python/sedonadb/python/sedonadb/datasource.py

```diff
@@ -0,0 +1,194 @@
+        columns or apply an output projection on an input stream.
+
+        The internals will keep a strong (Python) reference to the returned object
+        for as long as batches are being produced.
+
+        Args:
+            args: An object with attributes
+                - `src`: An object/file abstraction. Currently, `.to_url()` is the best way
+                  to extract the underlying URL from the source.
+                - `filter`: An object representing the filter expression that was pushed
+                  down, if one exists. Currently, `.bounding_box(column_index)` is the only
+                  way to interact with this object.
+                - `file_schema`: An optional schema. If `None`, the implementation must
+                  infer the schema.
+                - `file_projection`: An optional list of integers of the columns of
+                  `file_schema` that must be produced by this implementation (in the
+                  exact order specified).
+                - `batch_size`: An optional integer specifying the number of rows requested
+                  for each output batch.
+
+        """
+        raise NotImplementedError()
+
+    def infer_schema(self, src):
+        """Infer the output schema
+
+        Implementations can leave this unimplemented, in which case the internals will call
+        `open_reader()` and query the provided schema without pulling any batches.
+
+        Args:
+            src: An object/file abstraction. Currently, `.to_url()` is the best way
+                to extract the underlying URL from the source.
+        """
+        raise NotImplementedError()
+
+    def __sedona_external_format__(self):
+        return PyExternalFormat(self)
+
+
+class PyogrioFormatSpec(ExternalFormatSpec):
+    """An `ExternalFormatSpec` implementation wrapping GDAL/OGR via pyogrio"""
+
+    def __init__(self, extension=""):
+        self._extension = extension
+        self._options = {}
+
+    def with_options(self, options):
+        cloned = type(self)(self.extension)
+        cloned._options.update(options)
+        return cloned
+
+    @property
+    def extension(self) -> str:
+        return self._extension
+
+    def open_reader(self, args):
+        import pyogrio.raw
+
+        url = args.src.to_url()
+        if url is None:
+            raise ValueError(f"Can't convert {args.src} to OGR-openable object")
+
+        if url.startswith("http://") or url.startswith("https://"):
+            ogr_src = f"/vsicurl/{url}"
+        elif url.startswith("file://"):
+            ogr_src = url.removeprefix("file://")
+        else:
+            raise ValueError(f"Can't open {url} with OGR")
+
+        if ogr_src.endswith(".zip"):
+            ogr_src = f"/vsizip/{ogr_src}"
+
+        if args.is_projected():
+            file_names = args.file_schema.names
+            columns = [file_names[i] for i in args.file_projection]
```

Comment:
I believe that `file_names` are field/column names in the file schema. Maybe we can name it `columns`?